crates/server/src/verifysnsignature.rs

```diff
@@ -0,0 +1,313 @@
+use starknet_core::utils::get_selector_from_name;
+use starknet_crypto::{poseidon_hash_many, verify};
+use starknet_providers::jsonrpc::HttpTransport;
+use starknet_providers::{JsonRpcClient, Provider, Url};
+
+// Structure representing a field in a StarkNet type with a name and type
+#[derive(Serialize, Deserialize, Debug)]
+struct FieldType {
+    name: String,
+    #[serde(rename = "type")]
+    field_type: String,
+}
+
+// Structure holding definitions of StarkNet types: StarknetDomain and Challenge
+#[derive(Serialize, Deserialize, Debug)]
+struct Types {
+    #[serde(rename = "StarknetDomain")]
+    stark_net_domain: Vec<FieldType>,
+    #[serde(rename = "Challenge")]
+    challenge: Vec<FieldType>,
+}
+
+// Asynchronous function to verify an Argent wallet signature on chain
+pub async fn verify_argent_signature(
+    message_hash: String,
+    signature: Vec<String>,
+    wallet_address: String,
+    message: &str,
+    rpc_node_url: &str,
+    chain_id: &str,
+) -> eyre::Result<bool> {
+    // Convert inputs from strings to StarkNet-compatible types
+    let wallet_address = Felt::from_str(&wallet_address).unwrap();
+    let message_hash = Felt::from_str(&message_hash).unwrap();
+
+    // Set up a JSON-RPC client to interact with the StarkNet blockchain
+    let provider = JsonRpcClient::new(HttpTransport::new(Url::parse(rpc_node_url).unwrap()));
+
+    // Parse the signature strings into Felt types
+    let parsed_signature: Vec<_> = signature
+        .iter()
+        .map(|s| Felt::from_str(s).unwrap())
+        .collect();
+
+    // Formatted entry point selector (isValidSignature string) to needed format for StarkNet RPC call
+    let entry_point_selector: Felt =
+        Felt::from_str("0x213dfe25e2ca309c4d615a09cfc95fdb2fc7dc73fbcad12c450fe93b1f2ff9e")
+            .unwrap();
+
+    // Prepare a function call to verify the signature on-chain
+    let function_call = FunctionCall {
+        contract_address: wallet_address,
+        entry_point_selector,
+        calldata: {
+            let mut data = vec![message_hash, Felt::from(parsed_signature.len())];
+            data.extend_from_slice(&parsed_signature);
+            data
+        },
+    };
+
+    // Execute the function call on the latest block
+    let result = provider
+        .call(&function_call, BlockId::Tag(BlockTag::Latest))
+        .await;
+
+    match result {
+        // If the signature is valid, verify the hash
+        Ok(_) => {
+            let verify = verify_signature_hash(message_hash, wallet_address, message, chain_id);
+            if verify.is_ok() {
+                return Ok(true);
+            }
+            eyre::bail!("Invalid message hash");
+        }
+        Err(err) => {
+            eyre::bail!("Error verifying signature: {:?}", err);
+        }
+    }
+}
+
+// Function to verify a MetaMask Snap wallet signature off chain
+pub fn verify_metamask_signature(
+    message_hash: String,
```

Comment:
`&str` is a minimum-requirements `String`

if the function needs a `String` (which is an owned memory location) then it should reflect that in the function signature

and if it only needs a reference to a string, then `&str` is sufficient